File: integrations/app/src/main/java/app/revanced/bilibili/patches/CommentReplyUrlPatch.java
package app.revanced.bilibili.patches;

import androidx.annotation.Keep;

import com.bapis.bilibili.main.community.reply.v1.Content;

import app.revanced.bilibili.settings.Settings;

public class CommentReplyUrlPatch {
    @Keep
    public static void filterUrls(Content content) {
        if (!Settings.BlockWordSearch.get()) return;
        var urls = content.getMutableUrlsMap();
        var iterator = urls.entrySet().iterator();
        while (iterator.hasNext()) {
            var entry = iterator.next();
            var url = entry.getValue().getAppUrlSchema();
            if (url.startsWith("bilibili://search?from=appcommentline_search"))
                iterator.remove();
        }
    }
}

File: patches/src/main/kotlin/app/revanced/patches/bilibili/misc/protobuf/patch/CommentReplyUrlPatch.kt
package app.revanced.patches.bilibili.misc.protobuf.patch

import app.revanced.patcher.data.BytecodeContext
import app.revanced.patcher.extensions.InstructionExtensions.addInstruction
import app.revanced.patcher.patch.BytecodePatch
import app.revanced.patcher.patch.PatchException
import app.revanced.patcher.patch.annotation.CompatiblePackage
import app.revanced.patcher.patch.annotation.Patch

@Patch(
    name = "Comment word search url",
    description = "屏蔽评论关键词搜索功能",
    compatiblePackages = [
        CompatiblePackage(name = "tv.danmaku.bili"),
        CompatiblePackage(name = "tv.danmaku.bilibilihd"),
        CompatiblePackage(name = "com.bilibili.app.in")
    ]
)
object CommentReplyUrlPatch : BytecodePatch() {
    override fun execute(context: BytecodeContext) {
        context.findClass("Lcom/bapis/bilibili/main/community/reply/v1/Content;")
            ?.mutableClass?.methods?.find { it.name == "internalGetUrls" }?.addInstruction(
                0, """
                invoke-static {p0}, Lapp/revanced/bilibili/patches/CommentReplyUrlPatch;->filterUrls(Lcom/bapis/bilibili/main/community/reply/v1/Content;)V
            """.trimIndent()
            ) ?: throw PatchException("can not found internalGetUrls method")
    }
}
